internal/server/logbuffer/merge.go

```diff
@@ -0,0 +1,146 @@
+package logbuffer
+
+import (
+	"io"
+	"time"
+)
+
+// Merger can combine multiple log streams into one stream. It presumes
+// each stream emits entries in time order, and then weaves the entries
+// together to create a total time ordered stream.
+type Merger struct {
+	readers []MergeReader
+	heads   []TimedEntry
+}
+
+// TimedEntry is the interface each input returns entries in
+type TimedEntry interface {
+	// Time is the associated time of the value. This is used to sort
+	// the entry against other entries from other inputs.
+	Time() time.Time
+
+	// Value returns the log entry itself. This allows time addition
+	// wrappers to return their original value.
+	Value() interface{}
+}
+
+// MergeReader is value that returns TimedEntry's for Merger
+// to weave together.
+type MergeReader interface {
+	Next() (TimedEntry, error)
```

Comment:
Another Nitpick... `Next` feels a bit too vague. Perhaps `NextTimedEntry` here. Its a nitpick but the motivation is looking at the godoc for `Reader` it feels weird that Next returns a TimedEntry and Read doesn't. I think it'd be clearer to do this and make it easier for existing types to implement this (since `Next` might reasonably be implemented for OTHER things). 

Just a thought.

Reply:
Ah true, as an interface it's a bit to vague.